flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java

```diff
@@ -82,4 +91,78 @@ public static ResolveOrder fromString(String resolveOrder) {
 			super(urls, parent);
 		}
 	}
+
+	/**
+	 * Ensures that holding a reference on the context class loader outliving the scope of user code does not prevent
+	 * the user classloader to be garbage collected (FLINK-16245).
+	 *
+	 * <p>This classloader delegates to the actual user classloader. Upon {@link #close()}, the delegate is nulled
+	 * and can be garbage collected. Additional class resolution will be resolved solely through the bootstrap
+	 * classloader and most likely result in ClassNotFound exceptions.
+	 *
+	 * @param <T> the classloader type that also needs to be closeable.
+	 */
+	private static class SafetyNetWrapperClassLoader<T extends ClassLoader & Closeable> extends CloseableClassLoader
+			implements Closeable {
+		private static final Logger LOG = LoggerFactory.getLogger(SafetyNetWrapperClassLoader.class);
+
+		private T inner;
+
+		SafetyNetWrapperClassLoader(T inner) {
+			super(null);
+			this.inner = inner;
+		}
+
+		@Override
+		public void close() {
+			if (inner != null) {
+				try {
+					inner.close();
+				} catch (IOException e) {
+					LOG.warn("Could not close user classloader", e);
+				}
+			}
+			inner = null;
+		}
+
+		@Override
+		protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+			if (inner == null) {
+				return super.loadClass(name, resolve);
+			}
+
+			synchronized (getClassLoadingLock(name)) {
```

Comment:
why aren't we just forwarding the call to the inner class?

Reply:
`loadClass(String name, boolean resolve)` is protected, `loadClass(String name)` is not resolving by default.

Reply:
Since we only use it for our own custom classloaders we could modify the access level

Reply:
`inner` is typically a `URLClassLoader`. We could subclass it, but I'd actually like to avoid that unless you have a strong opinion on that.

Reply:
Given that `inner` is always a `Flink[Parent|Child]FirstClassLoader`, we could increase the visibility of the method.

Reply:
We would need to introduce a common ancestor to `Flink[Parent|Child]FirstClassLoader` that does nothing but exposes `loadClass`. 
I'm really not convinced that this is better than do the resolution as is. It would certainly not be more concise and imho it would increase the complexity.

Reply:
My guess is that there will be a common ancestor anyways int he future (to register disposal hooks, see [here](https://github.com/apache/flink/pull/11196#issuecomment-591347938)). But I am okay with having the in a follow-up effort.

Reply:
Added a common ancestor now and used delegation.